exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetScanBatchCreator.java

```diff
@@ -68,76 +83,131 @@ protected ScanBatch getBatch(ExecutorFragmentContext context, AbstractParquetRow
+              timer.start();
+            }
+
+            ParquetMetadata footer = readFooter(fs.getConf(), rowGroup.getPath(), readerConfig);
+            if (timer != null) {
+              long timeToRead = timer.elapsed(TimeUnit.MICROSECONDS);
+              logger.trace("ParquetTrace,Read Footer,{},{},{},{},{},{},{}", "", rowGroup.getPath(), "", 0, 0, 0, timeToRead);
+            }
+            footers.put(rowGroup.getPath(), footer);
           }
+          ParquetMetadata footer = footers.get(rowGroup.getPath());
+
+          //
+          //   If a filter is given (and it is not just "TRUE") - then use it to perform run-time pruning
+          //
+          if ( filterExpr != null && ! (filterExpr instanceof ValueExpressions.BooleanExpression)  ) { // skip when no filter or filter is TRUE
+
+            int rowGroupIndex = rowGroup.getRowGroupIndex();
+            long footerRowCount = footer.getBlocks().get(rowGroupIndex).getRowCount();
+
+            if ( timer != null ) {  // restart the timer, if tracing
+              timer.reset();
+              timer.start();
+            }
+
+            // When starting a new file, or at the first time - Initialize path specific metadata etc
+            if ( ! rowGroup.getPath().equals(prevRowGroupPath) ) {
+              // Get the table metadata (V3)
+              tableMetadataV3 = Metadata.getParquetTableMetadata(footer, fs, rowGroup.getPath().toString(), readerConfig);
+
+              // The file status for this file
+              FileStatus fileStatus = fs.getFileStatus(rowGroup.getPath());
+              List<FileStatus> listFileStatus = new ArrayList<>(Arrays.asList(fileStatus));
+              List<Path> listRowGroupPath = new ArrayList<>(Arrays.asList(rowGroup.getPath()));
+              List<ReadEntryWithPath> entries = new ArrayList<>(Arrays.asList(new ReadEntryWithPath(rowGroup.getPath())));
+              fileSelection = new FileSelection(listFileStatus, listRowGroupPath, selectionRoot);
+
+              metadataProvider = new ParquetTableMetadataProviderImpl(entries, selectionRoot, fileSelection.cacheFileRoot, readerConfig, fs,false);
+              // The file metadata (for all columns)
+              fileMetadataV3 = Metadata.getParquetFileMetadata_v3(tableMetadataV3, footer, fileStatus, fs, true, null, readerConfig);
+
+              prevRowGroupPath = rowGroup.getPath(); // for next time
+            }
+
+            MetadataBase.RowGroupMetadata rowGroupMetadata = fileMetadataV3.getRowGroups().get(rowGroup.getRowGroupIndex());
+
+            Map<SchemaPath, ColumnStatistics> columnsStatistics = ParquetTableMetadataUtils.getRowGroupColumnStatistics(tableMetadataV3, rowGroupMetadata);
+
+            List<SchemaPath> columns = columnsStatistics.keySet().stream().collect(Collectors.toList());
+
+            ParquetGroupScan parquetGroupScan = new ParquetGroupScan( context.getQueryUserName(), metadataProvider, fileSelection, columns, readerConfig, filterExpr);
```

Comment:
Creating a new instance of the ParquetGroupScan seems too much overhead since this is happening for each row group.   I understand how you may have arrived at this since the run-time does not have access to the planner objects such as the original ParquetGroupScan.   However, the only reason you need this is to create a FilterPredicate, which is accessible through the ParquetGroupScan.  Could we just create a single (dummy) instance of ParquetGroupScan using 1 file (or row group) outside the main For loop and just create the FilterPredicate right in the beginning before even iterating over all the row groups ?  The construction of the FilterPredicate in `getFilterPredicate()`  itself does not seem to have any direct dependence on any internal state of the ParquetGroupScan as far as I can tell.

Reply:
1. The total "per rowgroup" overhead (including the ParquetGroupScan construction) is usually less than a millisecond (on my Mac). (That does not include the "per file/footer" overhead).
2. In case the input list of rowgroups spans multiple files, we'd need to recreate a ParquetGroupScan for each file (could there be changes in the columns or schema between files ? Seem there is some dependence on those.)

Reply:
I am actually less concerned about the elapsed time and more about the number of objects created on the Java heap.  The `ParquetGroupScan` is at least 100 bytes each and since we are targeting use cases with about million row groups per minor fragment, it means 100 MB per minor fragment (assuming the garbage collector is not doing the cleanup while the Foo loop is executing).  If there are 24 minor frags, this would add up to 2.4GB heap per node which would cause quite a bit of GC activity subsequently.  

I have made some changes on top of your branch which avoids this by directly calling a static method to create a FilterPredicate.   The pruning unit tests are passing with these changes.  I will send my changes to you offline and we can discuss.